src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.Helpers.cs

```diff
@@ -9,103 +9,48 @@ namespace System.Text.Json
-        {
-            if (options == null)
-            {
-                options = JsonSerializerOptions.s_defaultOptions;
-            }
-
-            byte[] result;
-
-            using (var output = new PooledByteBufferWriter(options.DefaultBufferSize))
-            {
-                WriteCore(output, value, type, options);
-                result = output.WrittenMemory.ToArray();
-            }
-
-            return result;
-        }
-
-        private static string WriteCoreString(object? value, Type type, JsonSerializerOptions? options)
-        {
-            if (options == null)
-            {
-                options = JsonSerializerOptions.s_defaultOptions;
-            }
-
-            string result;
-
-            using (var output = new PooledByteBufferWriter(options.DefaultBufferSize))
-            {
-                WriteCore(output, value, type, options);
-                result = JsonReaderHelper.TranscodeHelper(output.WrittenMemory.Span);
-            }
-
-            return result;
-        }
-
-        private static void WriteValueCore(Utf8JsonWriter writer, object? value, Type type, JsonSerializerOptions? options)
-        {
-            if (options == null)
-            {
-                options = JsonSerializerOptions.s_defaultOptions;
-            }
+            Debug.Assert(writer != null);
 
-            if (writer == null)
+            //  We treat typeof(object) special and allow polymorphic behavior.
+            if (inputType == typeof(object) && value != null)
             {
-                throw new ArgumentNullException(nameof(writer));
+                inputType = value!.GetType();
             }
 
-            WriteCore(writer, value, type, options);
-        }
+            WriteStack state = default;
+            state.Initialize(inputType, options, supportContinuation: false);
+            JsonConverter jsonConverter = state.Current.JsonClassInfo!.PolicyProperty!.ConverterBase;
 
-        private static void WriteCore(PooledByteBufferWriter output, object? value, Type type, JsonSerializerOptions options)
-        {
-            using (var writer = new Utf8JsonWriter(output, options.GetWriterOptions()))
-            {
-                WriteCore(writer, value, type, options);
-            }
+            bool success = WriteCore(jsonConverter, writer, value, options, ref state);
+            Debug.Assert(success);
         }
 
-        private static void WriteCore(Utf8JsonWriter writer, object? value, Type type, JsonSerializerOptions options)
+        private static bool WriteCore<TValue>(
+            JsonConverter jsonConverter,
+            Utf8JsonWriter writer,
+            TValue value,
+            JsonSerializerOptions options,
+            ref WriteStack state)
         {
-            Debug.Assert(type != null || value == null);
             Debug.Assert(writer != null);
 
-            if (value == null)
```

Comment:
Before we were not calling Initialize when the root was `null`; now we do and that causes a change in behavior for unsupported root types.

```cs
[Fact]
public static void QuickTest()
{
    int[,] arr = null;
    string json = JsonSerializer.Serialize(arr);
    Console.WriteLine(json);
}
```

on master, the previous example returns null. with these changes we will throw.

Reply:
I think this can be fixed if you just keep this condition.

Reply:
Since we want to support converters writing their own `null` we must ask the converter if they want to handle null. For example a custom String converter may want to write `null` as `""`. So we just can't write `null` here.